ensime_shared/ensime.py

```diff
@@ -279,11 +336,15 @@ def au_buf_leave(self, client, filename):
 
     @execute_with_client()
     def au_cursor_hold(self, client, filename):
-        client.on_cursor_hold(filename)
+        """Handler for CursorHold autocommand event."""
+        self.watchdog_client_server_connection(client)
+        client.unqueue_and_display(filename)
+        client.editor.cursorhold()
```

Comment:
Thanks for testing @jvican. Would be curious about reproducible cases of blocking the UI if it's not completions, I think they're *probably* not results of this PR, but still.

In theory these changes could improve responsiveness: there's a subtle change where I moved the `time.sleep` in the polling thread's loop, it was running on every normal iteration and upon inspection this is needless, socket `recv` is efficient blocking I/O that's exactly what we want to happen with no artificial pause for this threading model. I moved the sleep to instead prevent a total busy-loop in the condition that we've lost the connection. This reconnection stuff isn't perfect yet but we'll focus on that soon; I'm not positive that there aren't race conditions (i.e. the `running` and `connected` attributes shared between threads without a queue), but these things haven't really changed here—a deep refactoring of this logic wasn't my intention, yet. I haven't actually been able to observe what I think is a race, the GIL may be solving it. If we're noticeably blocking Vim anywhere though, that's got to be the main thread.

I've been traveling this week so haven't had a chance to tie up the loose ends on this PR, will try this weekend.